File: packages/py/analytics/src/bolsa_analytics/cognitive/position_decision.py
"""PositionDecision — proyección operativa (V1.27). No durable, no segundo motor."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

from bolsa_analytics.cognitive.exit_plan import (
    ExitPlan,
    ExitPlanStatus,
    build_exit_plan_from_position,
)
from bolsa_analytics.cognitive.exit_policy import ExitPolicy, resolve_exit_policy
from bolsa_analytics.cognitive.position_state import PositionState

PositionDecisionAction = Literal[
    "HOLD", "PROTECT", "REDUCE", "TAKE_PROFIT", "EXIT", "REVIEW"
]
PositionAttention = Literal["NORMAL", "ATTENTION", "URGENT", "BLOCKED"]
PositionNextEvent = Literal[
    "NONE", "T1", "T2", "STOP", "TRAIL", "TIME", "THESIS_REVIEW", "RECONCILIATION"
]
PositionReconHealth = Literal["CLEAN", "ATTENTION", "CRITICAL"]
PositionProtection = Literal["ACTIVE", "NONE"]
PositionUrgency = Literal["LOW", "MEDIUM", "HIGH"]

POSITION_DECISION_KEY = "positionDecision"

_ATT_RANK = {"NORMAL": 0, "ATTENTION": 1, "URGENT": 2, "BLOCKED": 3}

_STATUS_CONFIDENCE: dict[ExitPlanStatus, float] = {
    "TRIGGERED": 0.9,
    "ARMED": 0.75,
    "HINT": 0.65,
    "IDLE": 0.55,
    "DONE": 0.5,
}

_RECON_CONFIDENCE: dict[PositionReconHealth, float] = {
    "CLEAN": 1.0,
    "ATTENTION": 0.85,
    "CRITICAL": 0.5,
}


def map_recon_status_to_health(status: str | None) -> PositionReconHealth:
    s = (status or "").strip().lower()
    if s == "drift":
        return "CRITICAL"
    if s in ("clean", "ok"):
        return "CLEAN"
    return "ATTENTION"


def recon_health_to_attention(health: PositionReconHealth) -> PositionAttention:
    if health == "CRITICAL":
        return "BLOCKED"
    if health == "ATTENTION":
        return "ATTENTION"
    return "NORMAL"


def attention_to_urgency(attention: PositionAttention) -> PositionUrgency:
    if attention in ("URGENT", "BLOCKED"):
        return "HIGH"
    if attention == "ATTENTION":
        return "MEDIUM"
    return "LOW"


#: Salidas PROTECTORAS: deshacer riesgo no espera a la reconciliación (invariante de oro
#: del roadmap AUTO-2 §6.4). Con el libro sin verificar, ``CRITICAL`` degrada a ``REVIEW``
#: **salvo** cuando el plan pide reducir riesgo ya (stop rebasado, trail alcanzado, riesgo
#: de cartera): en ese caso manda la salida. Cualquier otro motivo (objetivos, manual,
#: tesis, tiempo) sigue vetado: tomar beneficio puede esperar, no cubrirse no.
#: ``TIME_STOP`` NO entra (una salida por tiempo no es cubrirse) y ``THESIS_INVALIDATION``
#: tampoco (decisión D2 de 2b: vende cuando la reconciliación lo permite, y su veto bajo
#: ``CRITICAL`` se declara en vez de silenciarse).
_PROTECTIVE_EXIT_REASONS: frozenset[str] = frozenset(
    {"STRUCTURAL_STOP", "TRAIL", "PORTFOLIO_RISK"}
)


def _max_attention(a: PositionAttention, b: PositionAttention) -> PositionAttention:
    return a if _ATT_RANK[a] >= _ATT_RANK[b] else b


def _protection_state(position: PositionState) -> PositionProtection:
    if position.current_stop is not None:
        return "ACTIVE"
    return "NONE"


def _next_event(position: PositionState, exit_plan: ExitPlan) -> PositionNextEvent:
    primary = exit_plan.primary_reason
    if primary == "STRUCTURAL_STOP":
        return "STOP"
    if primary == "THESIS_INVALIDATION":
        return "THESIS_REVIEW"
    if primary == "TARGET_1":
        return "T1"
    if primary == "TARGET_2":
        return "T2"
    if primary == "TRAIL":
        return "TRAIL"
    if primary == "TIME_STOP":
        # V2.42 slice 2b (E1): el techo de mantenimiento es el motivo; que no se confunda
        # con "esperando T1/T2" (antes caía en la rama de objetivos).
        return "TIME"
    if not position.target1_achieved_at and position.target1 is not None:
        return "T1"
    if not position.target2_achieved_at and position.target2 is not None:
        return "T2"
    return "NONE"


def _mark_proximity_factor(
    position: PositionState,
    *,
    mark_price: float | None,
    next_event: PositionNextEvent,
) -> float:
    if mark_price is None or mark_price != mark_price:
        return 0.85
    target: float | None = None
    if next_event == "T1" and position.target1 is not None:
        target = position.target1
    elif next_event == "T2" and position.target2 is not None:
        target = position.target2
    if target is None or target <= 0:
        return 0.9
    entry = position.actual_entry or position.planned_entry
    if entry is None or entry <= 0:
        return 0.9
    span = abs(target - entry)
    if span <= 1e-9:
        return 0.9
    progress = abs(mark_price - entry) / span
    return min(1.0, max(0.75, 0.75 + progress * 0.25))


def _evidence_strength(
    exit_plan: ExitPlan,
    recon_health: PositionReconHealth,
    *,
    mark_price: float | None,
) -> float:
    score = 0.2
    if exit_plan.primary_reason:
        score += 0.25
    if mark_price is not None and mark_price == mark_price:
        score += 0.2
    if recon_health == "CLEAN":
        score += 0.2
    elif recon_health == "ATTENTION":
        score += 0.1
    if exit_plan.status in ("TRIGGERED", "ARMED"):
        score += 0.15
    return round(min(1.0, max(0.0, score)), 4)


def _decision_confidence(
    exit_plan: ExitPlan,
    recon_health: PositionReconHealth,
    position: PositionState,
    *,
    mark_price: float | None,
    next_event: PositionNextEvent,
) -> float:
    base = _STATUS_CONFIDENCE.get(exit_plan.status, 0.55)
    recon = _RECON_CONFIDENCE.get(recon_health, 0.85)
    proximity = _mark_proximity_factor(
        position, mark_price=mark_price, next_event=next_event
    )
    value = base * recon * proximity
    return round(min(1.0, max(0.0, value)), 4)


def _action_from_plan(
    exit_plan: ExitPlan,
    recon_health: PositionReconHealth,
    thesis_invalid: bool,
) -> PositionDecisionAction:
    # H-4/§6.4: una salida PROTECTORA no se veta por reconciliación; cualquier otro motivo
    # bajo ``CRITICAL`` degrada a ``REVIEW`` (tomar beneficio puede esperar).
    if (
        recon_health == "CRITICAL"
        and exit_plan.primary_reason not in _PROTECTIVE_EXIT_REASONS
    ):
        return "REVIEW"
    # V2.42 slice 2b (E3 · decisión D2 firmada por el owner): la invalidación CONFIRMADA
    # de la tesis es una salida REAL (proteger capital), no un ``REVIEW``. Antes esto
    # devolvía ``REVIEW`` ⇒ el AUTO declaraba la tesis rota y NO vendía: la posición
    # sobrevivía a su propia invalidación. ``THESIS_INVALIDATION`` está en
    # ``_HARD_TRIGGER`` (``exit_plan``), de modo que el plan ya pidió ``full_exit``; aquí
    # sólo se deja de vetar. Bajo ``CRITICAL`` sigue mandando la rama de arriba (la
    # invalidación NO está en ``_PROTECTIVE_EXIT_REASONS``): el veto de reconciliación
    # está declarado, no silenciado.
    if thesis_invalid or exit_plan.primary_reason == "THESIS_INVALIDATION":
        return "EXIT"
    sug = exit_plan.suggested_action
    primary = exit_plan.primary_reason
    if sug == "protect":
        return "PROTECT"
    if sug == "full_exit":
        return "EXIT"
    if sug == "reduce":
        if primary in ("TARGET_1", "TARGET_2"):
            return "TAKE_PROFIT"
        return "REDUCE"
    return "HOLD"


@dataclass(frozen=True, slots=True)
class PositionDecision:
    position_id: str
    trade_plan_id: str
    action: PositionDecisionAction
    reason: str
    confidence: float
    urgency: PositionUrgency
    evidence_strength: float
    attention: PositionAttention
    next_event: PositionNextEvent
    protection: PositionProtection
    recon_health: PositionReconHealth
    suggested_qty: float | None
    suggested_stop: float | None
    primary_reason: str | None
    market_as_of: str | None
    expires_at: str | None

    def to_dict(self) -> dict[str, Any]:
        return {
            "positionId": self.position_id,
            "tradePlanId": self.trade_plan_id,
            "action": self.action,
            "reason": self.reason,
            "confidence": self.confidence,
            "urgency": self.urgency,
            "evidenceStrength": self.evidence_strength,
            "attention": self.attention,
            "nextEvent": self.next_event,
            "protection": self.protection,
            "reconHealth": self.recon_health,
            "suggestedQty": self.suggested_qty,
            "suggestedStop": self.suggested_stop,
            "primaryReason": self.primary_reason,
            "marketAsOf": self.market_as_of,
            "expiresAt": self.expires_at,
        }


def build_position_decision(
    position: PositionState | None,
    *,
    mark_price: float | None = None,
    exit_policy: ExitPolicy | None = None,
    template_id: str | None = None,
    portfolio_recon_status: str | None = None,
    thesis_invalid: bool = False,
    at: str | None = None,
    now: str | None = None,
    expires_at: str | None = None,
    trail_hint: bool = False,
    trail_stop: float | None = None,
) -> PositionDecision | None:
    if position is None:
        return None
    if position.direction not in ("long", "short"):
        return None
    # AUTO-2: fuente ÚNICA de política. Antes ``template_id=None`` ⇒ ``policy=None`` ⇒ el
    # fallback suelto de ``suggestion_from_exit_policy`` (0.5/1.0) hacía que los dos
    # caminos AUTO cerraran T1 con fracciones distintas. ``resolve_exit_policy`` ya tiene
    # MODERATE como default declarado: se usa siempre, sin rama silenciosa.
    policy = exit_policy if exit_policy is not None else resolve_exit_policy(template_id)
    exit_plan = build_exit_plan_from_position(
        position,
        mark_price=mark_price,
        now=now,
        expires_at=expires_at,
        thesis_invalid=thesis_invalid,
        trail_hint=trail_hint,
        trail_stop=trail_stop,
        at=at,
        exit_policy=policy,
    )
    if exit_plan is None:
        return None

    recon_health = map_recon_status_to_health(portfolio_recon_status)
    attention = recon_health_to_attention(recon_health)
    action = _action_from_plan(exit_plan, recon_health, thesis_invalid)
    primary = exit_plan.primary_reason
    protection = _protection_state(position)

    if primary == "STRUCTURAL_STOP":
        attention = _max_attention(attention, "URGENT")
    elif thesis_invalid or primary == "THESIS_INVALIDATION":
        attention = _max_attention(attention, "URGENT")
    elif action in ("REDUCE", "TAKE_PROFIT", "EXIT", "PROTECT"):
        attention = _max_attention(attention, "ATTENTION")

    if recon_health == "CRITICAL":
        reason = "reconciliation:portfolio_drift"
        next_event: PositionNextEvent = "RECONCILIATION"
    elif primary:
        reason = primary.lower()
        next_event = _next_event(position, exit_plan)
    else:
        reason = "hold"
        next_event = _next_event(position, exit_plan)

    urgency = attention_to_urgency(attention)
    evidence_strength = _evidence_strength(
        exit_plan, recon_health, mark_price=mark_price
    )
    confidence = _decision_confidence(
        exit_plan,
        recon_health,
        position,
        mark_price=mark_price,
        next_event=next_event,
    )
    stamp = at or exit_plan.updated_at

    return PositionDecision(
        position_id=position.position_id,
        trade_plan_id=position.trade_plan_id,
        action=action,
        reason=reason,
        confidence=confidence,
        urgency=urgency,
        evidence_strength=evidence_strength,
        attention=attention,
        next_event=next_event,
        protection=protection,
        recon_health=recon_health,
        suggested_qty=exit_plan.suggested_qty,
        suggested_stop=exit_plan.suggested_stop,
        primary_reason=primary,
        market_as_of=stamp,
        expires_at=None,
    )

File: packages/shared/src/cognitive/position-decision.ts
/**
 * PositionDecision — proyección operativa de posición (V1.27).
 * No es entidad durable ni segundo motor de salida.
 * Consume ExitPlan (evento) + ExitPolicy + recon + tesis → qué hacer ahora.
 */

import {
  buildExitPlanFromPosition,
  type ExitPlanSignalsV1,
  type ExitPlanStatusV1,
  type ExitPlanV1,
} from "./exit-plan.js";
import { resolveExitPolicy, type ExitPolicyV1 } from "./exit-policy.js";
import type { PositionStateV1 } from "./position-state.js";

export type PositionDecisionActionV1 =
  | "HOLD"
  | "PROTECT"
  | "REDUCE"
  | "TAKE_PROFIT"
  | "EXIT"
  | "REVIEW";

export type PositionAttentionV1 = "NORMAL" | "ATTENTION" | "URGENT" | "BLOCKED";

export type PositionNextEventV1 =
  | "NONE"
  | "T1"
  | "T2"
  | "STOP"
  | "TRAIL"
  | "THESIS_REVIEW"
  | "RECONCILIATION";

export type PositionReconHealthV1 = "CLEAN" | "ATTENTION" | "CRITICAL";

export type PositionProtectionV1 = "ACTIVE" | "NONE";

export type PositionUrgencyV1 = "LOW" | "MEDIUM" | "HIGH";

export const RECON_HEALTH_COPY: Record<PositionReconHealthV1, string> = {
  CLEAN: "Operativa normal",
  ATTENTION: "Operativa: atención",
  CRITICAL: "Operativa bloqueada",
};

export type PositionDecisionV1 = {
  positionId: string;
  tradePlanId: string;
  action: PositionDecisionActionV1;
  reason: string;
  confidence: number;
  urgency: PositionUrgencyV1;
  evidenceStrength: number;
  attention: PositionAttentionV1;
  nextEvent: PositionNextEventV1;
  protection: PositionProtectionV1;
  reconHealth: PositionReconHealthV1;
  suggestedQty: number | null;
  suggestedStop: number | null;
  primaryReason: string | null;
  marketAsOf: string | null;
  expiresAt: string | null;
};

const STATUS_CONFIDENCE: Record<ExitPlanStatusV1, number> = {
  TRIGGERED: 0.9,
  ARMED: 0.75,
  HINT: 0.65,
  IDLE: 0.55,
  DONE: 0.5,
};

const RECON_CONFIDENCE: Record<PositionReconHealthV1, number> = {
  CLEAN: 1,
  ATTENTION: 0.85,
  CRITICAL: 0.5,
};

export function mapReconStatusToHealth(
  status: string | null | undefined,
): PositionReconHealthV1 {
  const s = String(status ?? "")
    .trim()
    .toLowerCase();
  if (s === "drift") return "CRITICAL";
  if (s === "clean" || s === "ok") return "CLEAN";
  return "ATTENTION";
}

export function reconHealthToAttention(
  health: PositionReconHealthV1,
): PositionAttentionV1 {
  if (health === "CRITICAL") return "BLOCKED";
  if (health === "ATTENTION") return "ATTENTION";
  return "NORMAL";
}

export function attentionToUrgency(
  attention: PositionAttentionV1,
): PositionUrgencyV1 {
  if (attention === "URGENT" || attention === "BLOCKED") return "HIGH";
  if (attention === "ATTENTION") return "MEDIUM";
  return "LOW";
}

function maxAttention(
  a: PositionAttentionV1,
  b: PositionAttentionV1,
): PositionAttentionV1 {
  const rank: Record<PositionAttentionV1, number> = {
    NORMAL: 0,
    ATTENTION: 1,
    URGENT: 2,
    BLOCKED: 3,
  };
  return rank[a] >= rank[b] ? a : b;
}

function protectionState(position: PositionStateV1): PositionProtectionV1 {
  return position.currentStop != null ? "ACTIVE" : "NONE";
}

function nextEventFromPosition(
  position: PositionStateV1,
  exitPlan: ExitPlanV1,
): PositionNextEventV1 {
  const primary = exitPlan.primaryReason;
  if (primary === "STRUCTURAL_STOP") return "STOP";
  if (primary === "THESIS_INVALIDATION") return "THESIS_REVIEW";
  if (primary === "TARGET_1") return "T1";
  if (primary === "TARGET_2") return "T2";
  if (primary === "TRAIL") return "TRAIL";
  if (!position.target1AchievedAt && position.target1 != null) return "T1";
  if (!position.target2AchievedAt && position.target2 != null) return "T2";
  return "NONE";
}

function markProximityFactor(
  position: PositionStateV1,
  markPrice: number | null | undefined,
  nextEvent: PositionNextEventV1,
): number {
  if (typeof markPrice !== "number" || !Number.isFinite(markPrice)) return 0.85;
  let target: number | null = null;
  if (nextEvent === "T1" && position.target1 != null) target = position.target1;
  else if (nextEvent === "T2" && position.target2 != null) {
    target = position.target2;
  }
  if (target == null || target <= 0) return 0.9;
  const entry = position.actualEntry ?? position.plannedEntry;
  if (entry == null || entry <= 0) return 0.9;
  const span = Math.abs(target - entry);
  if (span <= 1e-9) return 0.9;
  const progress = Math.abs(markPrice - entry) / span;
  return Math.min(1, Math.max(0.75, 0.75 + progress * 0.25));
}

function evidenceStrengthFromSignals(
  exitPlan: ExitPlanV1,
  reconHealth: PositionReconHealthV1,
  markPrice: number | null | undefined,
): number {
  let score = 0.2;
  if (exitPlan.primaryReason) score += 0.25;
  if (typeof markPrice === "number" && Number.isFinite(markPrice)) score += 0.2;
  if (reconHealth === "CLEAN") score += 0.2;
  else if (reconHealth === "ATTENTION") score += 0.1;
  if (exitPlan.status === "TRIGGERED" || exitPlan.status === "ARMED") {
    score += 0.15;
  }
  return Math.round(Math.min(1, Math.max(0, score)) * 10000) / 10000;
}

function decisionConfidence(
  exitPlan: ExitPlanV1,
  reconHealth: PositionReconHealthV1,
  position: PositionStateV1,
  markPrice: number | null | undefined,
  nextEvent: PositionNextEventV1,
): number {
  const base = STATUS_CONFIDENCE[exitPlan.status] ?? 0.55;
  const recon = RECON_CONFIDENCE[reconHealth];
  const proximity = markProximityFactor(position, markPrice, nextEvent);
  const value = base * recon * proximity;
  return Math.round(Math.min(1, Math.max(0, value)) * 10000) / 10000;
}

function actionFromPlan(
  exitPlan: ExitPlanV1,
  reconHealth: PositionReconHealthV1,
  thesisInvalid: boolean,
): PositionDecisionActionV1 {
  if (reconHealth === "CRITICAL") return "REVIEW";
  if (thesisInvalid || exitPlan.primaryReason === "THESIS_INVALIDATION") {
    return "REVIEW";
  }
  const primary = exitPlan.primaryReason;
  const sug = exitPlan.suggestedAction;
  if (sug === "protect") return "PROTECT";
  if (sug === "full_exit") return "EXIT";
  if (sug === "reduce") {
    if (primary === "TARGET_1" || primary === "TARGET_2") return "TAKE_PROFIT";
    return "REDUCE";
  }
  return "HOLD";
}

export type BuildPositionDecisionInputV1 = {
  position: PositionStateV1 | null | undefined;
  signals?: ExitPlanSignalsV1 | null;
  exitPolicy?: ExitPolicyV1 | null;
  templateId?: string | null;
  portfolioReconStatus?: string | null;
  thesisInvalid?: boolean | null;
  at?: string | null;
};

export function buildPositionDecision(
  input: BuildPositionDecisionInputV1,
): PositionDecisionV1 | null {
  const position = input.position;
  if (!position) return null;
  if (position.direction !== "long" && position.direction !== "short") {
    return null;
  }

  const policy =
    input.exitPolicy ??
    (input.templateId ? resolveExitPolicy(input.templateId) : undefined);
  const thesisInvalid = input.thesisInvalid === true;
  const signals: ExitPlanSignalsV1 = {
    ...(input.signals ?? {}),
    thesisInvalid: thesisInvalid || input.signals?.thesisInvalid === true,
    exitPolicy: policy ?? input.signals?.exitPolicy,
    at: input.at ?? input.signals?.at,
  };
  const exitPlan = buildExitPlanFromPosition(position, signals);
  if (!exitPlan) return null;

  const reconHealth = mapReconStatusToHealth(input.portfolioReconStatus);
  let attention = reconHealthToAttention(reconHealth);
  if (reconHealth === "CRITICAL") {
    attention = "BLOCKED";
  }

  const action = actionFromPlan(exitPlan, reconHealth, thesisInvalid);
  const primary = exitPlan.primaryReason;
  const protection = protectionState(position);

  if (primary === "STRUCTURAL_STOP") {
    attention = maxAttention(attention, "URGENT");
  } else if (thesisInvalid || primary === "THESIS_INVALIDATION") {
    attention = maxAttention(attention, "URGENT");
  } else if (
    action === "REDUCE" ||
    action === "TAKE_PROFIT" ||
    action === "EXIT" ||
    action === "PROTECT"
  ) {
    attention = maxAttention(attention, "ATTENTION");
  }

  let reason = "hold";
  if (reconHealth === "CRITICAL") reason = "reconciliation:portfolio_drift";
  else if (primary) reason = primary.toLowerCase();
  else if (action === "HOLD") reason = "hold";

  const nextEvent =
    reconHealth === "CRITICAL"
      ? "RECONCILIATION"
      : nextEventFromPosition(position, exitPlan);
  const markPrice = signals.markPrice;
  const urgency = attentionToUrgency(attention);
  const evidenceStrength = evidenceStrengthFromSignals(
    exitPlan,
    reconHealth,
    markPrice,
  );
  const confidence = decisionConfidence(
    exitPlan,
    reconHealth,
    position,
    markPrice,
    nextEvent,
  );

  const stamp =
    typeof input.at === "string" && input.at.trim()
      ? input.at
      : exitPlan.updatedAt;

  return {
    positionId: position.positionId,
    tradePlanId: position.tradePlanId,
    action,
    reason,
    confidence,
    urgency,
    evidenceStrength,
    attention,
    nextEvent,
    protection,
    reconHealth,
    suggestedQty: exitPlan.suggestedQty,
    suggestedStop: exitPlan.suggestedStop,
    primaryReason: exitPlan.primaryReason,
    marketAsOf: stamp,
    expiresAt: null,
  };
}
